presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java

```diff
@@ -0,0 +1,160 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.AggregationNode.Aggregation;
+import com.facebook.presto.sql.planner.plan.MarkDistinctNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.google.common.base.Predicates;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * Implements distinct aggregations with different inputs by transforming plans of the following shape:
+ * <pre>
+ * - Aggregation
+ *        GROUP BY (k),
+ *        F1(DISTINCT s0, s1, ...),
+ *        F2(DISTINCT s2, s3, ...)
+ *     - X
+ * </pre>
+ * into
+ * <pre>
+ * - Aggregation
+ *        F1(...) mask ($0)
+ *        F2(...) mask ($1)
+ *     - MarkDistinct (k, s0, s1, ...) -> $0
+ *          - MarkDistinct (k, s2, s3, ...) -> $1
+ *              - X
+ * </pre>
+ */
+public class MultipleDistinctAggregationToMarkDistinct
+        implements Rule<AggregationNode>
+{
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(Predicates.or(
+                    MultipleDistinctAggregationToMarkDistinct::hasMultipleDistinctInputsWithoutFilter,
+                    MultipleDistinctAggregationToMarkDistinct::hasMixedDistinctAndNonDistincts));
+
+    private static boolean hasMultipleDistinctInputsWithoutFilter(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations()
+                .values().stream()
+                .filter(e -> e.getCall().isDistinct() && !(e.getCall().getFilter().isPresent() || e.getMask().isPresent()))
+                .map(Aggregation::getCall)
+                .map(FunctionCall::getArguments)
+                .map(HashSet::new)
+                .distinct()
+                .count() > 1;
+    }
+
+    private static boolean hasMixedDistinctAndNonDistincts(AggregationNode aggregation)
+    {
+        long distincts = aggregation.getAggregations()
+                .values().stream()
+                .map(Aggregation::getCall)
+                .filter(FunctionCall::isDistinct)
+                .count();
+
+        return distincts > 0 && distincts < aggregation.getAggregations().size();
```

Comment:
why non distinct aggregation is required? Javadoc contradicts this predicate:
```
+ * Implements distinct aggregations with different inputs by transforming plans of the following shape:
+ * <pre>
+ * - Aggregation
+ *        GROUP BY (k),
+ *        F1(DISTINCT s0, s1, ...),
+ *        F2(DISTINCT s2, s3, ...)
```

Reply:
This handles the case where there’s at least one non-distinct agg. That is not something the other rule (single distinct to group by) can deal with

Reply:
> single distinct to group by

but for single distinct to group by all aggregation must have same symbols.

Reply:
Yes, so:

* if there are more than one different distincts OR
* there’s at least one non-distinct

This rule will handle it. I can update the example to reflect that

Reply:
ah right. Haven't seen `or` there. My bad.